### PR TITLE
extension: add test case

### PR DESCRIPTION
**Pull request by lambdai**
Commit Message:
Add extra test case to guard that cluster upstream extension support header rewrite.

Additional Description:

Risk Level: LOW
Testing: Integration test
Docs Changes:
Release Notes:
Platform Specific Features:

### CODE REVIEW

File: test/integration/cluster_upstream_extension_integration_test.cc [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  }
+  PerHostGenericConnPoolFactory per_host_upstream_factory_;
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, ClusterUpstreamExtensionIntegrationTest,
+                         testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                         TestUtility::ipTestParamsToString);
+
+TEST_P(ClusterUpstreamExtensionIntegrationTest,
+       VerifyRequestHeadersAreRewrittenByClusterAndHostMetadata) {
```
Comment (snowp) [inline test/integration/cluster_upstream_extension_integration_test.cc:61]: Could you add some comments here about what we're trying to verify here? Just to make it easier for future readers to grok what this test is about.

File: test/integration/upstreams/per_host_upstream_request.h [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        const auto& fields = data_struct.fields();
+        if (auto iter = fields.find("bar"); iter != fields.end()) {
+          if (iter->second.kind_case() == ProtobufWkt::Value::kStringValue) {
+            dup->setCopy(Envoy::Http::LowerCaseString("X-host-foo"), iter->second.string_value());
+          }
+        }
+      }
+    }
+    sub_upstream_.encodeHeaders(*dup, end_stream);
+  }
```
Comment (snowp) [inline test/integration/upstreams/per_host_upstream_request.h:95]: This reads very densely, could you break it up or add comments to make it clearer what's being done?

File: test/integration/upstreams/per_host_upstream_request.h [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    sub_upstream_.encodeTrailers(trailers);
+  }
+
+  void readDisable(bool disable) override { sub_upstream_.readDisable(disable); }
+
+  void resetStream() override { sub_upstream_.resetStream(); }
+
+private:
+  Extensions::Upstreams::Http::Http::HttpUpstream sub_upstream_;
+  Upstream::HostDescriptionConstSharedPtr host_{};
```
Comment (snowp) [inline test/integration/upstreams/per_host_upstream_request.h:107]: i would make this const and remove {} since we always initialize it in the ctor

File: test/integration/upstreams/per_host_upstream_request.cc [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+void PerHostHttpConnPool::newStream(GenericConnectionPoolCallbacks* callbacks) {
+  callbacks_ = callbacks;
+  // It's possible for a reset to happen inline within the newStream() call. In this case, we
+  // might get deleted inline as well. Only write the returned handle out if it is not nullptr to
+  // deal with this case.
+  Envoy::Http::ConnectionPool::Cancellable* handle =
+      conn_pool_->newStream(callbacks->upstreamToDownstream(), *this);
+  if (handle) {
+    conn_pool_stream_handle_ = handle;
+  }
```
Comment (snowp) [inline test/integration/upstreams/per_host_upstream_request.cc:18]: if check is redundant here, just do the assignment always

File: test/integration/upstreams/per_host_upstream_request.cc [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+using Envoy::Router::GenericConnectionPoolCallbacks;
+
+namespace Envoy {
+
+void PerHostHttpConnPool::newStream(GenericConnectionPoolCallbacks* callbacks) {
+  callbacks_ = callbacks;
+  // It's possible for a reset to happen inline within the newStream() call. In this case, we
+  // might get deleted inline as well. Only write the returned handle out if it is not nullptr to
+  // deal with this case.
```
Comment (snowp) [inline test/integration/upstreams/per_host_upstream_request.cc:13]: This comment doesn't really make sense to me since conn_pool_stream_handle_ is nullptr to begin with?

File: test/integration/upstreams/per_host_upstream_request.cc [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+absl::optional<Envoy::Http::Protocol> PerHostHttpConnPool::protocol() const {
+  return conn_pool_->protocol();
+}
+
+void PerHostHttpConnPool::onPoolFailure(ConnectionPool::PoolFailureReason reason,
+                                        absl::string_view transport_failure_reason,
+                                        Upstream::HostDescriptionConstSharedPtr host) {
+  callbacks_->onPoolFailure(reason, transport_failure_reason, host);
```
Comment (snowp) [inline test/integration/upstreams/per_host_upstream_request.cc:37]: presumably we should reset the handle ptr here as well?

File: test/integration/upstreams/per_host_upstream_config.cc [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+namespace Envoy {
+
+Router::GenericConnPoolPtr PerHostGenericConnPoolFactory::createGenericConnPool(
+    Upstream::ClusterManager& cm, bool is_connect, const Router::RouteEntry& route_entry,
+    absl::optional<Envoy::Http::Protocol> downstream_protocol,
+    Upstream::LoadBalancerContext* ctx) const {
+  if (is_connect) {
+    return nullptr;
+  }
+  auto ret =
```
Comment (snowp) [inline test/integration/upstreams/per_host_upstream_config.cc:14]: i would use a more descriptive name here, e.g. `conn_pool`

File: test/integration/upstreams/BUILD [hunk 8]
```diff
@@ -0,0 +1,52 @@
+load(
+    "//bazel:envoy_build_system.bzl",
+    "envoy_cc_extension",
+    "envoy_cc_library",
+    "envoy_extension_package",
+)
+
+licenses(["notice"])  # Apache 2
+
+#envoy_package()
```
Comment (snowp) [inline test/integration/upstreams/BUILD:10]: fix?

File: source/extensions/upstreams/http/http/BUILD [hunk 9]
```diff
@@ -33,6 +33,7 @@ envoy_cc_library(
     hdrs = [
         "upstream_request.h",
     ],
+    visibility = ["//visibility:public"],
```
Comment (snowp) [inline source/extensions/upstreams/http/http/BUILD:36]: Seems a bit odd for me to see cc_library with public visibility, why is this required?

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks for the test, I left a few comments to get this started

File: source/extensions/upstreams/http/http/BUILD (hunk 9 above)
Comment (lambdai) [reply source/extensions/upstreams/http/http/BUILD:36]: I think it's the instruction `envoy_extension_package()` turns the `upstream_request_lib` invisible to other packages. I can double check if I was doing anything wrong in the past.

Do you agree that upstream_request_lib could be utilized by other packages? I really need this upstream_request so that the upstream request could be wrapped in outer upstream request.

**Review by lambdai — COMMENTED** [review]
@snowp 

Thanks for the review! I will fix and update the PR later today

File: source/extensions/upstreams/http/http/BUILD (hunk 9 above)
Comment (lambdai) [reply source/extensions/upstreams/http/http/BUILD:36]: This is the error message if the visibility is removed.
```
$ bazel test //test/integration:cluster_upstream_extension_integration_test 
ERROR: /home/lambdai/workspace/envoy/test/integration/upstreams/BUILD:12:19: in cc_library rule //test/integration/upstreams:per_host_upstream_config: target '//source/extensions/upstreams/http/http:upstream_request_lib' is not visible from target '//test/integration/upstreams:per_host_upstream_config'. Check the visibility declaration of the former target if you think the dependency is legitimate
ERROR: Analysis of target '//test/integration:cluster_upstream_extension_integration_test' failed; build aborted: Analysis of target '//test/integration/upstreams:per_host_upstream_config' failed
```

File: test/integration/cluster_upstream_extension_integration_test.cc (hunk 1 above)
Comment (lambdai) [reply test/integration/cluster_upstream_extension_integration_test.cc:65]: add comment

File: test/integration/upstreams/BUILD (hunk 8 above)
Comment (lambdai) [reply test/integration/upstreams/BUILD]: fixed

File: test/integration/upstreams/per_host_upstream_config.cc (hunk 7 above)
Comment (lambdai) [reply test/integration/upstreams/per_host_upstream_config.cc]: fixed

File: test/integration/upstreams/per_host_upstream_request.cc (hunk 5 above)
Comment (lambdai) [reply test/integration/upstreams/per_host_upstream_request.cc]: fixed

File: test/integration/upstreams/per_host_upstream_request.cc (hunk 4 above)
Comment (lambdai) [reply test/integration/upstreams/per_host_upstream_request.cc]: good catch!

File: test/integration/upstreams/per_host_upstream_request.cc (hunk 6 above)
Comment (lambdai) [reply test/integration/upstreams/per_host_upstream_request.cc:31]: set   conn_pool_stream_handle_  to nullptr here and add comment to conn_pool_stream_handle_ declaration

File: test/integration/upstreams/per_host_upstream_request.h (hunk 2 above)
Comment (lambdai) [reply test/integration/upstreams/per_host_upstream_request.h]: add a static function to inject header

File: test/integration/upstreams/per_host_upstream_request.h (hunk 3 above)
Comment (lambdai) [reply test/integration/upstreams/per_host_upstream_request.h]: right. Fixed

**Review by lambdai — COMMENTED** [review]
@snowp I am seeing checkformat failure on my new "test/integration/upstream/BUILD"

The auto fix is to add envoy_package(), but that fix actually fail the entire build. Any suggestion?

**Comment by lambdai** [comment]
> @snowp I am seeing checkformat failure on my new "test/integration/upstream/BUILD"
> 
> The auto fix is to add envoy_package(), but that fix actually fail the entire build. Any suggestion?

The root cause: envoy_build_fixer.py likes either envoy_package or envoy_extension_package(sole in source/extensions).
The above make sense.

Now that the newly add extension is only used by integration test, I choose to declare in in `envoy_package`.

**Comment by snowp** [comment]
Seems like CI is broken, can you fix and then I'll take a look?

/wait

**Comment by lambdai** [comment]
@snowp Thank you for the reminder 
Yeah, I reproduce locally when I merged the latest master. Look like linux_x64 release is passing and I expect the rest will pass soon

File: test/integration/upstreams/per_host_upstream_request.h [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+  Http::Status encodeHeaders(const Envoy::Http::RequestHeaderMap& headers,
+                             bool end_stream) override {
+    auto dup = Envoy::Http::RequestHeaderMapImpl::create();
+    Envoy::Http::HeaderMapImpl::copyFrom(*dup, headers);
+    dup->setCopy(Envoy::Http::LowerCaseString("X-foo"), "foo-common");
+    addHeader(*dup, "X-cluster-foo", host_->cluster().metadata(), "foo", "bar");
+    if (host_->metadata() != nullptr) {
+      addHeader(*dup, "X-host-foo", *host_->metadata(), "foo", "bar");
+    }
```
Comment (snowp) [inline test/integration/upstreams/per_host_upstream_request.h:94]: If you're only trying to verify that you can modify headers in the upstream extension why the complexity here? Why not just a single header?

File: test/integration/cluster_upstream_extension_integration_test.cc [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, ClusterUpstreamExtensionIntegrationTest,
+                         testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                         TestUtility::ipTestParamsToString);
+
+// This test verify that cluster upstream extensions can fulfill the requirement that rewrite http
+// headers after cluster and host is selected. See https://github.com/envoyproxy/envoy/issues/12236
+// This test case should be rewritten once upstream http
+// filters(https://github.com/envoyproxy/envoy/issues/10455) is landed.
```
Comment (snowp) [inline test/integration/cluster_upstream_extension_integration_test.cc:63]: If this is something that we want to support and validate via a test it seems like we should keep this test for perpetuity - just because you'd rewrite your extension to be a HTTP filter instead it doesn't mean everyone else would.

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks this looks better, just a few comments.

Would appreciate @alyssawilk's thoughts as well, in particular on the general usefulness of this test: part of me feels like this is pretty guaranteed to continue to work no matter what happens in the router and elsewhere.

File: test/integration/upstreams/per_host_upstream_request.h (hunk 10 above)
Comment (lambdai) [reply test/integration/upstreams/per_host_upstream_request.h:94]: The idea is that I don't want to add test case because 
"Envoy can do this in this current implementation" but  
"This is real world requirement"
Also this can acts an example for user who want to rewrite header base on per upstream attribute like #13897

Does it make sense?

File: test/integration/cluster_upstream_extension_integration_test.cc (hunk 11 above)
Comment (lambdai) [reply test/integration/cluster_upstream_extension_integration_test.cc:63]: As I explained in the other comment: This is an unfortunate constraint。 
I am happy to refactor my upstream extension if upstream HTTP filter could achieve the goal providing alternative api.
So consider this as soft requirement when you are working on it

**Comment by lambdai** [comment]
gentle ping @alyssawilk

File: test/integration/cluster_upstream_extension_integration_test.cc [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+#include "envoy/registry/registry.h"
+#include "envoy/router/router.h"
+
+#include "common/buffer/buffer_impl.h"
+
+#include "test/integration/http_integration.h"
+#include "test/integration/upstreams/per_host_upstream_config.h"
+#include "test/test_common/registry.h"
+
+#include "fake_upstream.h"
```
Comment (alyssawilk) [inline test/integration/cluster_upstream_extension_integration_test.cc:12]: I think generally we do full paths, so test/integration/fake_upstream.h"

File: test/integration/cluster_upstream_extension_integration_test.cc [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+namespace {
+class ClusterUpstreamExtensionIntegrationTest
+    : public testing::TestWithParam<Network::Address::IpVersion>,
+      public HttpIntegrationTest {
+public:
+  ClusterUpstreamExtensionIntegrationTest()
+      : HttpIntegrationTest(Http::CodecClient::Type::HTTP1, GetParam()) {}
+
+  void SetUp() override {
```
Comment (alyssawilk) [inline test/integration/cluster_upstream_extension_integration_test.cc:25]: I think you can remove this function - the defaults are HTTP/1 and the client is explicitly set above.

File: test/integration/cluster_upstream_extension_integration_test.cc [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+    setDownstreamProtocol(Http::CodecClient::Type::HTTP1);
+    setUpstreamProtocol(FakeHttpConnection::Type::HTTP1);
+  }
+
+  void populateMetadataTestData(envoy::config::core::v3::Metadata& metadata, const std::string& k1,
```
Comment (alyssawilk) [inline test/integration/cluster_upstream_extension_integration_test.cc:30]: key1 key2?

File: test/integration/cluster_upstream_extension_integration_test.cc [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+    HttpIntegrationTest::initialize();
+  }
+  PerHostGenericConnPoolFactory per_host_upstream_factory_;
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, ClusterUpstreamExtensionIntegrationTest,
+                         testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                         TestUtility::ipTestParamsToString);
+
+// This test verify that cluster upstream extensions can fulfill the requirement that rewrite http
```
Comment (alyssawilk) [inline test/integration/cluster_upstream_extension_integration_test.cc:60]: verify -> verifies
that -> that they

File: test/integration/cluster_upstream_extension_integration_test.cc [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+// headers after cluster and host is selected. See https://github.com/envoyproxy/envoy/issues/12236
+// This test case should be rewritten once upstream http
+// filters(https://github.com/envoyproxy/envoy/issues/10455) is landed.
+TEST_P(ClusterUpstreamExtensionIntegrationTest,
+       VerifyRequestHeadersAreRewrittenByClusterAndHostMetadata) {
+  initialize();
+  Registry::InjectFactory<Router::GenericConnPoolFactory> registration(per_host_upstream_factory_);
+
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+  auto response = codec_client_->makeHeaderOnlyRequest(default_request_headers_);
```
Comment (alyssawilk) [inline test/integration/cluster_upstream_extension_integration_test.cc:70]: sendRequestAndWaitForResponse() ?

File: test/integration/upstreams/per_host_upstream_config.cc [hunk 17]
```diff
@@ -0,0 +1,20 @@
+#include "test/integration/upstreams/per_host_upstream_config.h"
+
+#include "test/integration/upstreams/per_host_upstream_request.h"
+
+namespace Envoy {
+
+Router::GenericConnPoolPtr PerHostGenericConnPoolFactory::createGenericConnPool(
```
Comment (alyssawilk) [inline test/integration/upstreams/per_host_upstream_config.cc:7]: can we get away with .cc files only for this and below, as we do with test/integration/filters?

Most of the factory-only code doesn't need header files but I can't recall if this does.

Also as I think upstreams/config will be 1:1 I think we can stick them in the same file to make it easier to reason about and have one top-of-file comment of what it's for.

File: test/integration/upstreams/per_host_upstream_request.h [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+    if (auto iter = fields.find(meta_key_1); iter != fields.end()) {
+      if (iter->second.kind_case() == ProtobufWkt::Value::kStringValue) {
+        header_map.setCopy(Envoy::Http::LowerCaseString(std::string(header_name)),
+                           iter->second.string_value());
+      }
+    }
+  }
+}
+} // namespace
+class PerHostHttpConnPool : public Router::GenericConnPool,
```
Comment (alyssawilk) [inline test/integration/upstreams/per_host_upstream_request.h:35]: Can this inherit from HttpConnPool and only override functions where different work is needed?

**Review by alyssawilk — COMMENTED** [review]
I agree that it feels clear that this functionality will stick around.  I will say given how often Envoy APIs shift around it isn't terrible to have a canonical example, so that if API / factory changes happen there's also a canonical transformation PR, if that makes sense?  Basically don't know if it's necessary but I'm not averse.

Either way, sorry for the lag, and here's some thoughts on  how we can limit code duplication and tighten it up a bit :-)

File: test/integration/upstreams/per_host_upstream_request.h (hunk 18 above)
Comment (lambdai) [reply test/integration/upstreams/per_host_upstream_request.h]: Yeah, I made it

File: test/integration/upstreams/per_host_upstream_config.cc (hunk 17 above)
Comment (lambdai) [reply test/integration/upstreams/per_host_upstream_config.cc]: merge config.h and upstream_request.h

**Review by lambdai — COMMENTED** [review]
Thank you @alyssawilk 

Now the extension is a single header file. It is much cleaner for the test.

File: test/integration/upstreams/per_host_upstream_config.h [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+  void encodeTrailers(const Envoy::Http::RequestTrailerMap& trailers) override {
+    sub_upstream_.encodeTrailers(trailers);
+  }
+
+  void readDisable(bool disable) override { sub_upstream_.readDisable(disable); }
+
+  void resetStream() override { sub_upstream_.resetStream(); }
+
+private:
+  Extensions::Upstreams::Http::Http::HttpUpstream sub_upstream_;
```
Comment (alyssawilk) [inline test/integration/upstreams/per_host_upstream_config.h:73]: Oh, I was thinking if you inherit from the HTTP upstream rather than wrap it, you can remove all the wrapper functions (which simply call sub_upstream_.function).  WDYT?

File: test/integration/cluster_upstream_extension_integration_test.cc [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+  }
+  PerHostGenericConnPoolFactory per_host_upstream_factory_;
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, ClusterUpstreamExtensionIntegrationTest,
+                         testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                         TestUtility::ipTestParamsToString);
+
+// This test verifies that cluster upstream extensions can fulfill the requirement that they rewrite
+// http headers after cluster and host is selected. See
```
Comment (alyssawilk) [inline test/integration/cluster_upstream_extension_integration_test.cc:56]: is selected -> are selected

**Review by alyssawilk — COMMENTED** [review]
Cool, much cleaner, thanks!
Just two more comments and you're good to go.

File: test/integration/cluster_upstream_extension_integration_test.cc (hunk 20 above)
Comment (lambdai) [reply test/integration/cluster_upstream_extension_integration_test.cc]: fixed

File: test/integration/upstreams/per_host_upstream_config.h (hunk 19 above)
Comment (lambdai) [reply test/integration/upstreams/per_host_upstream_config.h]: Updated. The new shape is less error prone for this task. Thanks!

**Review by alyssawilk — APPROVED** [review]

**Review by snowp — APPROVED** [review]
